File: jlox-in-cpp/Interpreter.cpp
#include "Interpreter.h"

#include <iostream>
#include <sstream>

#include "Error.h"
#include "RuntimeError.h"
#include "Token.h"

void Interpreter::interpret(Expr expr) {
  try {
    Value value = std::visit(*this, expr);
    std::cout << stringify(value) << "\n";
  } catch (const RuntimeError &error) {
    runtimeError(error);
  }
}

auto Interpreter::operator()(const LiteralExpr *expr) -> Value {
  return std::visit([](auto &&v) -> Value { return v; }, expr->value);
}

auto Interpreter::operator()(const GroupingExpr *expr) -> Value {
  return std::visit(*this, expr->expr);
}

auto Interpreter::operator()(const UnaryExpr *expr) -> Value {
  Value right = std::visit(*this, expr->right);

  switch (expr->op.type) {
  case TokenType::BANG:
    return !isTruthy(right);

  case TokenType::MINUS:
    checkNumberOperand(expr->op, right);
    return -std::get<double>(right);

  default:
    __builtin_unreachable();
  }
}

auto Interpreter::operator()(const BinaryExpr *expr) -> Value {
  Value left = std::visit(*this, expr->left);
  Value right = std::visit(*this, expr->right);

  switch (expr->op.type) {
  case TokenType::BANG_EQUAL:
    return left != right;

  case TokenType::EQUAL_EQUAL:
    return left == right;

  case TokenType::GREATER:
    checkNumberOperands(expr->op, left, right);
    return std::get<double>(left) > std::get<double>(right);

  case TokenType::GREATER_EQUAL:
    checkNumberOperands(expr->op, left, right);
    return std::get<double>(left) >= std::get<double>(right);

  case TokenType::LESS:
    checkNumberOperands(expr->op, left, right);
    return std::get<double>(left) < std::get<double>(right);

  case TokenType::LESS_EQUAL:
    checkNumberOperands(expr->op, left, right);
    return std::get<double>(left) <= std::get<double>(right);

  case TokenType::MINUS:
    checkNumberOperands(expr->op, left, right);
    return std::get<double>(left) - std::get<double>(right);

  case TokenType::PLUS:
    if (std::holds_alternative<double>(left) &&
        std::holds_alternative<double>(right))
      return std::get<double>(left) + std::get<double>(right);

    if (std::holds_alternative<std::string_view>(left) &&
        std::holds_alternative<std::string_view>(right))
      return saveString(std::string(std::get<std::string_view>(left)) +
                        std::string(std::get<std::string_view>(right)));

    throw RuntimeError(expr->op,
                       "Operands must be two numbers or two strings.");

  case TokenType::SLASH:
    checkNumberOperands(expr->op, left, right);
    return std::get<double>(left) / std::get<double>(right);

  case TokenType::STAR:
    checkNumberOperands(expr->op, left, right);
    return std::get<double>(left) * std::get<double>(right);

  default:
    __builtin_unreachable();
  }
}

bool Interpreter::isTruthy(Value value) {
  if (std::holds_alternative<std::nullptr_t>(value))
    return false;
  if (bool *b = std::get_if<bool>(&value))
    return *b;
  return true;
}

void Interpreter::checkNumberOperand(const Token &token, Value value) {
  if (!std::holds_alternative<double>(value))
    throw new RuntimeError(token, "Operand must be a number.");
}

void Interpreter::checkNumberOperands(const Token &token, Value left,
                                      Value right) {
  if (!std::holds_alternative<double>(left) ||
      !std::holds_alternative<double>(right))
    throw new RuntimeError(token, "Operands must be numbers.");
}

std::string Interpreter::stringify(Value value) {
  // This is identical to AstPrinter's LiteralExpr logic right now, but it might
  // not be in the future, so duplicate it for now.
  static struct {
    std::string operator()(double d) {
      // Going this route instead of directly using std::string to avoid
      // unnecessary decimal places.
      std::ostringstream stream;
      stream << d;
      return stream.str();
    }

    std::string operator()(std::string_view s) { return std::string(s); }
    std::string operator()(bool b) { return b ? "true" : "false"; }
    std::string operator()(std::nullptr_t) { return "nil"; }
  } visitor;
  return std::visit(visitor, value);
}

std::string_view Interpreter::saveString(std::string &&str) {
  savedStrings.push_back(std::move(str));
  return savedStrings.back();
}

File: jlox-in-rust/src/interpreter.rs
use crate::error;
use crate::expr::{Expr, Literal};
use crate::token::UnaryTokenType;
use crate::value::{Value, ValueError};

pub struct Interpreter {
    // This will gain some fields later.
}

struct RuntimeError {
    message: &'static str,
    line: u32,
}

impl Interpreter {
    pub fn new() -> Self {
        Self {}
    }

    pub fn interpret(&self, expr: &Expr) {
        let result = self.evaluate(expr);
        match result {
            Ok(value) => println!("{value}"),
            Err(RuntimeError { message, line }) => error::runtime_error(message, line),
        }
    }

    // We'll use self later
    #[allow(clippy::only_used_in_recursion)]
    fn evaluate(&self, expr: &Expr) -> Result<Value, RuntimeError> {
        match expr {
            Expr::Literal(literal) => match *literal {
                Literal::Nil => Ok(Value::Nil),
                Literal::Boolean(boolean) => Ok(Value::Boolean(boolean)),
                Literal::Number(number) => Ok(Value::Number(number)),
                Literal::String(string) => Ok(Value::String(string.into())),
            },

            Expr::Grouping(expr) => self.evaluate(expr),

            Expr::Unary(token, right) => {
                let right = self.evaluate(right)?;
                let result = match token.r#type {
                    UnaryTokenType::Bang => !&right,
                    UnaryTokenType::Minus => -&right,
                };
                result.map_err(|ValueError(message)| RuntimeError {
                    message,
                    line: token.line,
                })
            }

            Expr::Binary(left, token, right) => {
                let left = self.evaluate(left)?;
                let right = self.evaluate(right)?;
                let result = match token.r#type {
                    crate::token::BinaryTokenType::Minus => &left - &right,
                    crate::token::BinaryTokenType::Plus => &left + &right,
                    crate::token::BinaryTokenType::Slash => &left / &right,
                    crate::token::BinaryTokenType::Star => &left * &right,
                    crate::token::BinaryTokenType::BangEqual => left.ne(&right),
                    crate::token::BinaryTokenType::EqualEqual => left.eq(&right),
                    crate::token::BinaryTokenType::Greater => left.gt(&right),
                    crate::token::BinaryTokenType::GreaterEqual => left.ge(&right),
                    crate::token::BinaryTokenType::Less => left.lt(&right),
                    crate::token::BinaryTokenType::LessEqual => left.le(&right),
                };
                result.map_err(|ValueError(message)| RuntimeError {
                    message,
                    line: token.line,
                })
            }
        }
    }
}
